freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaRecommendationService.java

```diff
@@ -0,0 +1,84 @@
+package com.sequenceiq.freeipa.service.stack;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.Maps;
+import com.sequenceiq.cloudbreak.cloud.model.CloudVmTypes;
+import com.sequenceiq.cloudbreak.cloud.model.VmType;
+import com.sequenceiq.cloudbreak.cloud.model.VmTypeMeta;
+import com.sequenceiq.cloudbreak.cloud.service.CloudParameterService;
+import com.sequenceiq.common.api.type.CdpResourceType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.create.FreeIpaRecommendationResponse;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.create.VmTypeResponse;
+import com.sequenceiq.freeipa.converter.cloud.CredentialToExtendedCloudCredentialConverter;
+import com.sequenceiq.freeipa.converter.instance.VmTypeToVmTypeResponseConverter;
+import com.sequenceiq.freeipa.dto.Credential;
+import com.sequenceiq.freeipa.service.CredentialService;
+import com.sequenceiq.freeipa.service.stack.instance.DefaultInstanceTypeProvider;
+
+@Service
+public class FreeIpaRecommendationService {
+
+    @Inject
+    private CredentialService credentialService;
+
+    @Inject
+    private CloudParameterService cloudParameterService;
+
+    @Inject
+    private CredentialToExtendedCloudCredentialConverter extendedCloudCredentialConverter;
+
+    @Inject
+    private DefaultInstanceTypeProvider defaultInstanceTypeProvider;
+
+    @Inject
+    private VmTypeToVmTypeResponseConverter vmTypeConverter;
+
+    public FreeIpaRecommendationResponse getRecommendation(String credentialCrn, String region, String availabilityZone) {
+        Credential credential = credentialService.getCredentialByCredCrn(credentialCrn);
+        String defaultInstanceType = defaultInstanceTypeProvider.getForPlatform(credential.getCloudPlatform());
+        Set<VmTypeResponse> availableVmTypes = getAvailableVmTypes(region, availabilityZone, credential, defaultInstanceType);
+        return new FreeIpaRecommendationResponse(availableVmTypes, defaultInstanceType);
+    }
+
+    private Set<VmTypeResponse> getAvailableVmTypes(String region, String availabilityZone, Credential credential, String defaultInstanceType) {
+        CloudVmTypes vmTypes = cloudParameterService.getVmTypesV2(extendedCloudCredentialConverter.convert(credential),
+                region, credential.getCloudPlatform(), CdpResourceType.DEFAULT, Maps.newHashMap());
+
+        Set<VmType> availableVmTypes = Collections.emptySet();
+        if (StringUtils.isNotBlank(availabilityZone)) {
+            availableVmTypes = vmTypes.getCloudVmResponses().get(availabilityZone);
```

Comment:
Is it possible to encounter a `null` result here, thus leading to an NPE below in line 63 when traversing `availableVmTypes.stream()`?

Reply:
The same logic is used in core during Datahub creation, thus I would say its safe to use.